### ISSUE TITLE
Refactor LTI role handling 

### ISSUE THREAD

**Issue by marcospri** [hypothesis/lms#4749]
TLDR, with the following plan the role handling in LMS should be saner:

- [x] Add scope as a concept in LTIRole
- [x] Unify `LTIUser.roles` to use `LTIRole`
- [x] Take scope into account when checking for `is_instructor/learner`

### How roles currently work on LMS

We use  roles from the LTI launch in two contexts:

#### Authorization

Based on the roles from the launch we decide authorization (can you edit assignments? can you grade them?)

In LTI1.1 we look at the raw string from the launch and do very basic checking for instructor/learner:

https://github.com/hypothesis/lms/blob/b32b97a308e0e7cb8a6127fd80dde705683183f6/lms/models/lti_user.py#L33

In LTI1.3, where we are very often get sent roles in institution and system scopes we first process the roles. We exclude non-context roles as those are not relevant in the course.

https://github.com/hypothesis/lms/blob/b32b97a308e0e7cb8a6127fd80dde705683183f6/lms/models/lti_params.py

Once that filtering is done they arrive at the same `is_instructor`|`is_learner` functions that only look at strings.

#### Data gathering

We often want to store the relationship between an user and "something" (a course, assignment...) in those cases we created a new LTIRole model that stores the raw role from the launch and a mapping to a known set of roles (admin, instructor and learner).

https://github.com/hypothesis/lms/blob/b32b97a308e0e7cb8a6127fd80dde705683183f6/lms/models/lti_role.py

We then use that model in `_membership` tables like:

https://github.com/hypothesis/lms/blob/b32b97a308e0e7cb8a6127fd80dde705683183f6/lms/models/assignment_membership.py#L29

### Current issues

In LTI roles are scoped to one of system, institution or context (course).

In 1.3 roles are almost always scoped while in 1.1 only sometimes.

When we first released 1.3 I was not aware of this feature and we continue parsing the roles as usual causing issues because for example an "instructor" role with the institution scope meant that the user was a teacher in some course but not in the current one (the only one our app cares about).

The fix for this was to ignore institution roles in LTI1.3 

- https://github.com/hypothesis/lms/pull/4373
 
We never tackled this issue in 1.1 so we probably have some weird edge cases bugs due to this since the beginning.

- https://github.com/hypothesis/lms/issues/4411

Now we have reports of admin users not being able to create courses when they are only admins (but not instructors in the course) 

- https://github.com/hypothesis/lms/issues/4702

This is due to us now ignoring all the system and institution roles.

The fix for this is to parse the scope part of the role and use it in our application code to have more fine grained control over access.

### Suggested plan of action

- Add the new, nullable column `scope` to the roles table
- Add code that fills the new column
- Review the data in the table after the initial fill and a few days worth of data in the wild (unlikely to get many new roles if any)
- Switch to the new logic based on LTIRole instead of strings.
- Add a new "user_role" table 
- Remove the existing "role" column in "user".

**Comment by marcospri** [comment]
This is mostly done except for the DB schema side of things. Tracking that on: https://github.com/hypothesis/lms/issues/3639
